pkg/utils/revision/revision_utils.go

```diff
@@ -0,0 +1,368 @@
+func ControllerRevisionName(prefix string, hash string) string {
+	if len(prefix) > 223 {
+		prefix = prefix[:223]
+	}
+
+	return fmt.Sprintf("%s-%s", prefix, hash)
+}
+
+// NewControllerRevision returns a ControllerRevision with a ControllerRef pointing to parent and indicating that
+// parent is of parentKind. The ControllerRevision has labels matching template labels, contains Data equal to data, and
+// has a Revision equal to revision. If the returned error is nil, the returned ControllerRevision is valid. If the
+// returned error is not nil, the returned ControllerRevision is invalid for use.
+func NewControllerRevision(parent metav1.Object,
+	parentKind schema.GroupVersionKind,
+	templateLabels map[string]string,
+	data runtime.RawExtension,
+	revision int64) *appsv1.ControllerRevision {
+	labelMap := make(map[string]string)
+	for k, v := range templateLabels {
+		labelMap[k] = v
+	}
+	cr := &appsv1.ControllerRevision{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:          labelMap,
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(parent, parentKind)},
+			Namespace:       parent.GetNamespace(),
+		},
+		Data:     data,
+		Revision: revision,
+	}
+	hash := HashControllerRevision(cr)
+	cr.Name = ControllerRevisionName(parent.GetName(), hash)
+	if cr.Labels[leaderworkerset.TemplateRevisionHashKey] == "" {
+		cr.Labels[leaderworkerset.TemplateRevisionHashKey] = hash
+	}
+	return cr
+}
+
+// HashControllerRevision hashes the contents of revision's Data using FNV hashing.
+// The returned hash will be a safe encoded string to avoid bad words.
+func HashControllerRevision(revision *appsv1.ControllerRevision) string {
+	hf := fnv.New32()
+	if len(revision.Data.Raw) > 0 {
+		hf.Write(revision.Data.Raw)
+	}
+	if revision.Data.Object != nil {
+		DeepHashObject(hf, revision.Data.Object)
+	}
+	return rand.SafeEncodeString(fmt.Sprint(hf.Sum32()))
+}
+
+// EqualRevision returns true if lhs and rhs are either both nil, if the templateRevisionHash is the same,
+// or if they are semantically equivalent.
+func EqualRevision(lhs *appsv1.ControllerRevision, rhs *appsv1.ControllerRevision) bool {
+
+	if lhs == nil || rhs == nil {
+		return lhs == rhs
+	}
+
+	if lhs.Labels[leaderworkerset.TemplateRevisionHashKey] == rhs.Labels[leaderworkerset.TemplateRevisionHashKey] {
+		return true
+	}
+
+	return bytes.Equal(lhs.Data.Raw, rhs.Data.Raw) && apiequality.Semantic.DeepEqual(lhs.Data.Object, rhs.Data.Object)
+}
+
+type History struct {
+	client.Client
+	context context.Context
+}
+
+// ListControllerRevisions lists all ControllerRevisions matching selector and owned by parent or no other
+// controller. If the returned error is nil the returned slice of ControllerRevisions is valid. If the
+// returned error is not nil, the returned slice is not valid.
+func (h *History) ListControllerRevisions(parent metav1.Object, selector labels.Selector) ([]*appsv1.ControllerRevision, error) {
```

Comment:
Just pass in the context and client and remove the struct. I would also rename it to ListRevisions to match all the other functions in the file.